gtsam/linear/NoiseModel.cpp

```diff
@@ -287,36 +284,41 @@ Diagonal::Diagonal(const Vector& sigmas)
 Diagonal::shared_ptr Diagonal::Variances(const Vector& variances, bool smart) {
   if (smart) {
     // check whether all the same entry
-    size_t n = variances.size();
-    for (size_t j = 1; j < n; j++)
-      if (variances(j) != variances(0)) goto full;
-    return Isotropic::Variance(n, variances(0), true);
+    if ((variances.array() == variances(0)).all()) {
```

Comment:
goto no longer make sense here (even if does elsewhere). Replace with 
```
return (smart && ( == )) ? Isotropic::Variance(...) : std::make_shared<Diagonal>(...);
```

Also, while at it, use `std::make_shared` wherever we can in this file.

Reply:
We can't use `make_shared` here since the constructor is protected. :(